lib/common/rules.c

```diff
@@ -1420,31 +1434,32 @@ pcmk_evaluate_rule(xmlNode *rule, const pcmk_rule_input_t *rule_input,
             break;
     }
 
-    crm_trace("Testing rule %s", pcmk__xe_id(rule));
-    for (expr = pcmk__xe_first_child(rule, NULL, NULL, NULL); expr != NULL;
-         expr = pcmk__xe_next(expr)) {
-
-        test = (pcmk__evaluate_condition(expr, rule_input,
-                                         next_change) == pcmk_rc_ok);
-        empty = FALSE;
+    // Evaluate each condition
+    for (xmlNode *expr = pcmk__xe_first_child(rule, NULL, NULL, NULL);
+         expr != NULL; expr = pcmk__xe_next(expr)) {
 
-        if (test && (combine == pcmk__combine_or)) {
-            crm_trace("Expression %s/%s passed",
-                      pcmk__xe_id(rule), pcmk__xe_id(expr));
-            return pcmk_rc_ok;
-
-        } else if (!test && (combine == pcmk__combine_and)) {
-            crm_trace("Expression %s/%s failed",
-                      pcmk__xe_id(rule), pcmk__xe_id(expr));
-            return pcmk_rc_op_unsatisfied;
+        empty = false;
+        if (pcmk__evaluate_condition(expr, rule_input,
+                                     next_change) == pcmk_rc_ok) {
+            if (combine == pcmk__combine_or) {
+                rc = pcmk_rc_ok; // Any pass is final for "or"
+                break;
+            }
+        } else if (combine == pcmk__combine_and) {
+            rc = pcmk_rc_op_unsatisfied; // Any failure is final for "and"
```

Comment:
Do we want to keep the return value from `pcmk__evaluate_condition()` so that we can return whatever error the evaluator returns? Doesn't matter much and I can see merit to both ways

Reply:
I don't think so, because the return value applies only to the single expression. If the rule has multiple, the message could be confusing.